# Applied Computational Genomics Course at UU: Spring 2017
- Faculty: Aaron Quinlan (aquinlan at genetics.utah.edu)
- Teaching assistants: Julie Fusier, Jingtao Guo, and Tom Sasani 
- Meets Tu and Th from 9:10-10:30 in HSEB 2948
- January 10 - April 25
- 2 credit hours

# Overview
This course will provide a comprehensive introduction to fundamental concepts and experimental approaches in the analysis and interpretation of experimental genomics data. It will be structured as a series of lectures covering key concepts and analytical strategies. A diverse range of biological questions enabled by modern DNA sequencing technologies will be explored including sequence alignment, the identification of genetic variation, structural variation, and ChIP-seq and RNA-seq analysis. Students will learn and apply the fundamental data formats and analysis strategies that underlie computational genomics research. The primary goal of the course is for students to be grounded in theory and have the ability to conduct independent genomic analyses.

# Prerequisites
- Online introduction to Linux. Students must complete the following two online tutorials before class begins. If they don't things will be very complicated very quickly.
  - https://www.codecademy.com/en/courses/learn-the-command-line/lessons/environment/exercises/bash-profile
  - http://rik.smith-unna.com/command_line_bootcamp/?id=9xnbkx6eaof
- An Apple computer or a windows machine with Putty installed.

# Other ideas
- Lecture on the history of computational biology - with a focus on genomics.

# Syllabus
- **Lecture 1 (Tu Jan 10; Quinlan): Course overview and Intro to UNIX**
    - **Required** Reading Prior to Lecture: 
        - Part 1 of [Unix and Perl Primer for Biologists](http://korflab.ucdavis.edu/Unix_and_Perl/current.pdf)
    - Topics covered
        - Brief history of computational biology
        - Course computing environment
        - Intro. to UNIX: Part 1
            - Logging in
            - The "shell"
            - Navigation
            - File system
            - Files
            - Basic commands
- **Lecture 2 (Th Jan 12; Quinlan): Intro to UNIX Part 2**
    - **Required** Reading Prior to Lecture: 
        - Part 2 (Advanced UNIX) of [Unix and Perl Primer for Biologists](http://korflab.ucdavis.edu/Unix_and_Perl/current.pdf)
    - Topics covered
        - Intro. to UNIX: Part 2
    - **Homework 1 assigned. (due by start of class, Jan 17)**
- **Lecture 3 (Tu Jan 17; Quinlan): What is a genome?**
    - **Required** Reading Prior to Lecture: 
        - [Initial sequencing and analysis of the human genome](http://www.nature.com/nature/journal/v409/n6822/full/409860a0.html)
    - Topics covered
      - How was the genome sequenced?
        - sequencing technology
        - assembly strategy
      - Chromosomes
        - size
        - gene content
        - centromeres
          - alpha satellites
      - Haplotypes
      - Genes and transcripts
        - how many?
        - isoforms
      - Repeat content
        - mobile elements
        - simple repeats
        - microsatellites
        - segmental duplication
      - GC content, banding
      - CpG islands  
- **Lecture 4 (Th Jan 19; Quinlan): Using UNIX to find patterns in a genome**
    - **Required** Reading Prior to Lecture:    
        - None.     
    - Topics covered
      - Thought experiment: kmers and kmer uniqueness.
      - Restriction enzymes
      - Origins of replication
      - Basic pattern matching using:
        - head
        - grep
        - cut
        - sort
        - uniq
      - Basic regular expressions
      - Unix pipes
    - **Homework 2 assigned (due Jan 24)**
- **Lecture 5 (Tu Jan 24; Quinlan): Mutation, recombination and genetic variation**
    - **Required** Reading Prior to Lecture: 
        - [A global reference for human genetic variation](http://www.nature.com/nature/journal/v526/n7571/full/nature15393.html)
    - Topics covered
      - Mutation and mechanisms
      - Recombination
      - Polymorphism
      - SNPs
      - Landscape of human genetic variation
          - Alleles and genotypes
          - Allele frequency spectrum
          - Hardy weinberg equilibrium
          - Population stratification and Fixation index
      - INDELs
      - SV and CNV
      - Human mutation rates
- **Lecture 6 (Th Jan 26; Quinlan): Modern DNA sequencing technologies**
    - **Required** Reading Prior to Lecture: 
        - [Coming of age: ten years of next-generation sequencing technologies](http://www.nature.com/nrg/journal/v17/n6/full/nrg.2016.49.html)
    - Topics covered
        - Sanger dideoxy sequencing
        - Illumina sequencing
            - Overview of technology
            - Paired-end vs. single-end
        - Pacbio
        - Oxford nanopore
        - Base calling
        - FASTQ format
    - **Homework 3 (working with the FASTQ format) assigned (due Jan 31)**
- **Lecture 7 (Tu Jan 31; Quinlan): DNA sequence alignment**
    - **Required** Reading Prior to Lecture: 
        - [Alignment of Next-Generation Sequencing Reads](http://www.annualreviews.org/doi/abs/10.1146/annurev-genom-090413-025358?journalCode=genom)
    - Topics covered
      - Sequence alignment
          - Theory
          - Mapping versus alignment
          - Local versus global alignment
              - Smith waterman
              - Needleman-wunsch
          - Basic challenges with modern DNA sequences
          - Tools
      - The SAM/BAM format
      - samtools and IGV
    - **Homework 4 (creating and working with SAM/BAM files with samtools and IGV) assigned (due Feb 2)**
- **Lecture 8 (Th Feb 2; Quinlan and/or Marth): Detecting genetic variation, part 1**
    - **Required** Reading Prior to Lecture: 
        - [A framework for variation discovery and genotyping using next-generation DNA sequencing data](http://www.nature.com/ng/journal/v43/n5/full/ng.806.html)
    - **Optional** Reading Prior to Lecture: 
        - [A general approach to single-nucleotide polymorphism discovery.](http://www.nature.com/ng/journal/v23/n4/full/ng1299_452.html)
    - Topics covered
      - Polymorphism rate, nucleotide diversity, heterozygosity
          - variation between species
          - variation regionally within a species.
      - SNP and INDEL calling
          - Theory
              - Basic concept
              - Sequencing error
              - Bayes theorem and priors
      - Assigning a genotype
      - Common problems and artifacts
          - paralogy
          - low depth
          - high error rate
          - ambiguous alignment
      - Single sample detection
- **Lecture 9 (Tu Feb 7; Quinlan): Detecting genetic variation, part 2**
    - **Required** Reading Prior to Lecture: 
        - [The VCF format and VCFtools](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3137218/)
    - Topics covered
        - VCF format
          - Attributes
          - Genotypes
        - Population calling
        - Basic annotations
    - Exploring the format
        - examples
        - IGV
    - Manipulating VCF with bcftools
    - **Homework 5 (variant calling and working with VCF files with bcftools and UNIX) assigned (due Feb 9)**
- **Lecture 10 (Th Feb 9; Quinlan): Exome sequencing in studies of human disease**
    - **Required** Reading Prior to Lecture: 
        - [Exome sequencing as a tool for Mendelian disease gene discovery](http://www.nature.com/nrg/journal/v12/n11/full/nrg3031.html)
    - Topics covered
        - Exome capture
        - Variant detection in families
          - Mendelian inheritance
          - Mendelian violations (de novo mutations, LoH, uniparental disomy)
        - The Exome Aggregation Consortium
        - Mendelian disease analysis
            - Variant annotation
            - Using population allele frequency
    - **Homework 6 (variant annotation and prioritization with SnpEff) assigned (due Feb 14)**
- **Lecture 11 (Tu Feb 14; Quinlan): Annotating of genetic variation continued**
    - **Required** Reading Prior to Lecture: 
        - [Choice of transcripts and software has a large effect on variant annotation](https://genomemedicine.biomedcentral.com/articles/10.1186/gm543)
    - Topics covered
        - Concepts
          - e.g, synonymous, non-synonymous
          - frameshift
          - stopgain
          - constraint
          - impact of transcript model
    - Tools
        - Polyphen
        - CADD
        - VEP
        - SnpEff
    - Homework (due Feb 2)
- **Lecture 12 (Th Feb 16; Quinlan): Annotating a genome**
    - **Required** Reading Prior to Lecture: 
        - None
    - Topics covered
      - How and why do we annotate a genome?
      - Conservation
      - CpG islands
      - Repeatmasker
      - Chromatin modifications
      - DNA methylations
      - Linkage blocks
  - **Homework 7 (exploring genome annotation files with UNIX) assigned (due Feb 21)**
- **Lecture 13 (Tu Feb 21; Quinlan): Genome data formats and genome arithmetic**
    - **Required** Reading Prior to Lecture: 
        - None
    - Topics covered
      - The genome as a coordinate system
      - BED format
      - GFF format
      - VCF format
      - Recap BAM format
      - UCSC and Biomart to retrieve genome annotations
      - UCSC and IGV to visualize 
- **Lecture 14 (Th Feb 23; Quinlan): Applied genome arithmetic with bedtools; part 1**
    - **Required** Reading Prior to Lecture: 
        - [BEDTools: the Swiss‐army tool for genome feature analysis](http://onlinelibrary.wiley.com/doi/10.1002/0471250953.bi1112s47/abstract?userIsAuthenticated=false&deniedAccessCustomisedMessage=)
    - Topics covered
      - The genome as a coordinate system revisited
      - Basic concepts of genome arithmetic
      - Introduction to bedtools
    - **Homework 8 (basic genome arithmetic with bedtools) assigned (due Feb 28)**
- **Lecture 15 (Tu Feb 28; Quinlan): Applied genome arithmetic with bedtools; part 2**
- **Lecture 16 (Th Mar 2; Quinlan): Digging deeper into UNIX**
    - pipes
    - basic scripts and pipelines
    - basic math and stats on the command line
    - Homework (due Mar 7)
- **Lecture 17 (Tu Mar 7; Quinlan): ChIP-seq principles**
    - experimental design
    - protocols
    - examples
- **Lecture 18 (Th Mar 9; Quinlan): ChIP-seq analysis**
    - analyses
    - toolsets
    - Homework (due Mar 14)
- **Lecture 19 (Tu Mar 14; Quinlan): RNA-seq principles**
    - experimental design
    - protocols
    - examples
- **Lecture 20 (Th Mar 16; Quinlan): RNA-seq analysis**
    - analyses
    - toolsets
    - Homework (due Mar 21)
- **Lecture 21 (Tu Mar 21; Quinlan): Basic probability**
    - Probability with coins and dice
    - Probability with DNA
    - Conditional probabilities
    - Use R for examples
    - Homework (due Mar 23)
- **Lecture 22 (Th Mar 23; Quinlan): Probability distributions**
    - Binomial
    - Gaussian
    - Poisson
    - Applications
- **Lecture 23 (Tu Mar 28; Quinlan): Data visualization, part 1**
    - Why
    - Pattern recognition
    - Detect problems
    - Ansombe’s quartet
- **Lecture 24 (Th Mar 30; Quinlan): Data visualization, part 2**
    - http://www.nature.com/collections/qghhqm/pointsofsignificance
    - Scatter plots
    - Histograms
    - Box whiskers
- **Lecture 25 (Tu Apr 4; Quinlan): Class project introduction: identifying and characterizing de novo mutations in genomes**
    - work in pairs
    - de novo mutation and human disease
    - Project (due Apr 25)
        - introduce problem and dataset
            - dataset with 4 families of 4
            - they must find the mutation causing a different disease phenotype in each family
- **Lecture 26 (Th Apr 6; Quinlan): Class project**
    - Introduction to gemini
    - Hands on
- **Lecture 27 (Tu Apr 11; Quinlan): Class project**
    - Gemini de novo tutorial, Q and A
- **Lecture 28 (Th Apr 13; Quinlan): Class project**
    - OMIM
    - ClinVar
    - ExAC
- **Lecture 29 (Tu Apr 18; Quinlan): Class project**
    - Common artifacts in identifying de novo mutations
    - Mutation versus polymorphism
- **Lecture 29 (Th Apr 20; Quinlan): Class project**
    - Questions, group discussion
- **Lecture 30 (Tu Apr 25; Quinlan): Final class**
    - Group presentations
    - Class evaluations
